torch/_inductor/triton_ops/autotune.py

```diff
@@ -429,14 +431,18 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
     raise NotImplementedError(f"size_hints: {size_hints}")
 
 
-def reduction(size_hints, reduction_hint=False, meta=None, filename=None):
+def reduction(
+    size_hints, reduction_hint=False, max_regs=2**14, meta=None, filename=None
+):
     """args to @triton.heuristics()"""
     assert meta is not None
     rnumel = size_hints[-1]
     if len(size_hints) == 2:
+        u = 2 ** (max_regs.bit_length() - 1)  # previous power of 2
         contiguous_config = triton_config_reduction(
-            size_hints, 1, (rnumel if 256 <= rnumel < 2048 else 2048), num_stages=1
+            size_hints, 1, max(min(rnumel, u), 256), num_stages=1, max_warps=32
```

Comment:
I don't like 32 warps (it's easy to run into "insufficient resources" at launch with 32 warps), I also don't understand what's the conceptual meaning of `u`

Reply:
u is the upper bound for the `RBLOCK` size. We could call it `MAX_RBLOCK`?
I have not hit insufficient resources when running torchbench. Do you have an example? Note that the number of warps is dependent on the size of `RBLOCK`. In particular, here, since we are passing `x=1` and `r=clamp(rnumel, 32, 2**14)` (or about `2**14`) `RBLOCK` be about the size of `rnumel`. So, going through the logic in `triton_config_reduction`, we get that for `32 <= rnumel <= 16k (=2**14)` (for most reduction), we set `XNUMEL = 1`, `RNUMEL = rnumel` and `warps = clamp(RNUMEL // 128, 2, 32)`. This gives the following regimes:
- A bit of under utilisement in the range `rnumel = 32` and `rnumel = 256` (as it was before). Did not see any performance benefits increasing the min warps to 1 really.
- An unrolling of 4 and full warp engagement between `RNUMEL = 256, num_warps = 2` and `RNUMEL = 32*128 = 4096`. Same performance as before.
- Increase up to an unrolling of 16 up to `RNUMEL = 16k`. In the softmax case, the rough estimate of registry use per thread that we computed before would give us that we would be hitting 64 registers per thread (bit too much) but in reality, after the SMs clever allocations, the profiler shows that it's able to do it in 50 registers.
- After that, stay with an unrolling of 16 and some iterations of the for loop in triton up to `rnumel = 512k`, where we start splitting.

We see in the benchmarks a generous speed-up in the last two regimes.

Reply:
This is pretty close to what @Chillee had in his initial experiments, where he just increased the RBLOCK to `rnumel`, and yet apparently that proved pretty bad on the whole torchbench suite. @chillee do you have something saved from those experiments?